sysbuild: generate .config.sysbuild for controlling build settings

tejlmand opened the pull request:
Sysbuild now generates a .config.sysbuild config file which specifies
settings controlled by sysbuild.
Any setting specified in this .config will overrule user provided
setting, and a warning will be raised if the sysbuild controlled value
is different from the value specified by the user.

This has the following benefits:
- Allow sysbuild to control any build setting without adjustments to
the existing Kconfig tree
- Allow sysbuild to adjust settings based on knowledge regarding enabled
images / bootloaders.
- Cleanup CMake code, as settings in sysbuild no longer needs to be
propagated using CMake cache variables.

-----

See alternative approach here: #55616 

----

Pros in https://github.com/zephyrproject-rtos/zephyr/pull/55871

- No changes required in existing Kconfig tree.
- Sysbuild controlled settings always wins
- Easier to introduce a new sysbuild controlled setting, as it can simply be done by a single command, like: `set_config_bool(${image} CONFIG_<name> <y|n>)`

Cons

- Less visible to users when a given setting is controlled by sysbuild

------

A combination of #55871 and #55616 can be done, so that some symbols could have their title adjusted so that when sysbuild controls the setting, then the setting is still visible, but title could be like:
```
[*] MCUboot bootloader support (sysbuild controlled)
```

Fixes #53126

tejlmand committed:
kconfig: provide an option for enforcing Kconfig settings

This commit introduces an internal FORCE_CONF_FILE CMake setting which
allows higher order build systems to generate a configuration file
which will always take precedence.

This means that in case a user tries to change any setting to be
different than the defined value in the FORCE_CONF_FILE provide file(s),
then a warning will be printed and the setting will be reset to the
value given in the FORCE_CONF_FILE file.

Example of such warning:
   <path>/.config.sysbuild:1: warning: BOOTLOADER_MCUBOOT
   (defined at Kconfig.zephyr:766) set more than once.
   Old value "n", new value "y".

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
sysbuild: generate .config.sysbuild for controlling build settings

Sysbuild now generates a .config.sysbuild config file which specifies
settings controlled by sysbuild.
Any setting specified in this .config will overrule user provided
setting, and a warning will be raised if the sysbuild controlled value
is different from the value specified by the user.

This has the following benefits:
- Allow sysbuild to control any build setting without adjustments to
  the existing Kconfig tree
- Allow sysbuild to adjust settings based on knowledge regarding enabled
  images / bootloaders.
- Cleanup CMake code, as settings in sysbuild no longer needs to be
  propagated using CMake cache variables.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

nordicjm commented on cmake/modules/kconfig.cmake:
```diff
@@ -297,10 +297,14 @@ if(EXISTS ${DOTCONFIG} AND EXISTS ${merge_config_files_checksum_file})
 endif()
```
Typos in commit:
`This commit introducuced an internal CONFIG_FORCED CMake setting which` -> `This commit introduces an internal CONFIG_FORCED CMake setting which`
`allows higher order build systems to generated a configuration file` -> `allows higher order build systems to generate a configuration file`
`which will always take precendence` -> `which will always take precedence`

tejlmand replied:
done

nordicjm reviewed (changes requested):
Inside the mcuboot .config.sysbuild file:
```
CONFIG_BOOT_SIGNATURE_KEY_FILE="/tmp/aa/bootloader/mcuboot/root-rsa-2048.pem"
CONFIG_MCUBOOT_GENERATE_UNSIGNED_IMAGE=n
```
The second one is not valid for mcuboot, it is for applications only

57300 commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -321,16 +342,25 @@ function(ExternalZephyrProject_Cmake)
                    ${${ZCMAKE_APPLICATION}_CACHE_FILE} ONLY_IF_DIFFERENT
   )
 
+  set(dotconfigsysbuild ${${ZCMAKE_APPLICATION}_KCONFIG_BINARY_DIR}/.config.sysbuild)
```
nit: I expected to find `.config.sysbuild` this next to `${app}/zephyr/.config`, not in `${app}/Kconfig`.

tejlmand replied:
Good point.
Other Kconfig generated files usually goes to `Kconfig` dir, but I agree, the name `.config.sysbuild` could indicate to users that this file ought to be located next to `.config`.

I will relocate, but allow others to give their feedback before doing so.

tejlmand replied:
Moved to `${app}/zephyr/.config.sysbuild`

57300 commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -321,16 +342,25 @@ function(ExternalZephyrProject_Cmake)
+  set(dotconfigsysbuild ${${ZCMAKE_APPLICATION}_KCONFIG_BINARY_DIR}/.config.sysbuild)
+  get_target_property(config_content ${ZCMAKE_APPLICATION} CONFIG)
+  string(CONFIGURE "${config_content}" config_content)
+  file(WRITE ${dotconfigsysbuild} ${config_content})
```
Writes `config_content-NOTFOUND` if empty.

tejlmand replied:
nice catch, guess it happened as side-effect on the board revision issue.

Will ensure the `CONFIG` setting is always initialized with a value.

tejlmand replied:
Fixed.

57300 commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -382,3 +412,15 @@ function(sysbuild_module_call)
+
+function(set_config_bool image setting value)
+  if(${value})
+    set_property(TARGET ${image} APPEND_STRING PROPERTY CONFIG "${setting}=y\n")
```
What is the advantage of doing it this way, instead of using CMake cache?
```
set(${image}_${setting} "y" CACHE STRING "" FORCE)
```

57300 commented on share/sysbuild/image_config.cmake:
```diff
@@ -0,0 +1,23 @@
+# This sysbuild CMake file sets the sysbuild controlled settings as properties
+# on Zephyr images.
+
+get_target_property(image_board ${image} BOARD)
+if((NOT image_board) OR ("${image_BOARD}" STREQUAL "${BOARD}"))
```
This doesn't work for boards which have a default revision.
For example, if I build for `nrf9160dk_nrf9160`, then: `image_board == "nrf9160dk_nrf9160@0.7.0" != BOARD`

tejlmand replied:
good catch 👍 
This happens because the BOARD here is always without revision because [boards.cmake](https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/boards.cmake) has split the setting, whereas the `image_BOARD` is the raw value.

Will take a look.

57300 replied:
This is still an issue.

One possible solution is to cherry-pick https://github.com/zephyrproject-rtos/zephyr/pull/58174/commits/fa1a137003cfadc54211573ef8dd5afb757a16fa, if you'd like (from my sysbuild grouping PR).

tejlmand replied:
> This is still an issue (doesn't work for boards with a default revision).

I know, which is why I didn't mark it as fixed 😉 
When looking at it I discovered another flaw wrt. BOARD that I wanted to address first.
Part of it which you have also discovered, but there is also the possibility of a user specifying a dedicated BOARD for an image, such as `-DBOARD=myboard -Dmcuboot_BOARD=myboard@<revision>` where user wants to build mcuboot for specific revision even though the main app could be targeting another revision.

In such a case, the revision will be part of the image `BOARD` setting.

(edit: comment has been moved)

57300 commented on share/sysbuild/image_config.cmake:
```diff
@@ -0,0 +1,23 @@
+get_target_property(image_board ${image} BOARD)
+if((NOT image_board) OR ("${image_BOARD}" STREQUAL "${BOARD}"))
+  get_target_property(${image}_APP_TYPE ${image} APP_TYPE)
+  if(NOT ${image}_APP_TYPE STREQUAL BOOTLOADER)
+    set_config_bool(${image} CONFIG_BOOTLOADER_MCUBOOT "${SB_CONFIG_BOOTLOADER_MCUBOOT}")
+    set_config_string(${image} CONFIG_MCUBOOT_SIGNATURE_KEY_FILE
+                               "${SB_CONFIG_BOOT_SIGNATURE_KEY_FILE}"
+    )
+  endif()
```
Looking at this, `APP_TYPE` seems like something that should be extended to a more generic `GROUP` property. I think it would be useful to group images arbitrarily and set a common configuration for all images in (or not in) a given group.

It would also be nice to let one image belong to multiple groups and, eventually, to assign existing images to new groups defined out of tree.

tejlmand replied:
sounds like a good idea.

Will take a look.

tejlmand replied:
covered by this comment: https://github.com/zephyrproject-rtos/zephyr/pull/55871#discussion_r1202951853

57300 reviewed (commented):
> @57300 thanks for the review of the implementation itself.
> 
> Do you have any comments regarding the approach itself, that is the creation of a `.config.sysbuild` which overrules the settings from `.config` ? Feel free to comment also in relation to the alternative in #55616.

Sure! I like the `.config.sysbuild` approach overall. Most importantly, it seems to guarantee that sysbuild-controlled settings are always correctly propagated and can never be overridden at image level, and I can't think of a better way to do that. What's also nice is that, by applying this file on top of `.config`, we can make changes with `ninja sysbuild_menuconfig` without wiping out all changes from `ninja menuconfig`.

I also think the `set_config_xxxx()` functions look pretty tidy. The Kconfig `@template@` files in #55616 are probably more expressive in some sense, but here it looks easier to control individual settings for particular images without potentially having to generate a lot of Kconfig files.

You've noted that the downside of this approach is poor visibility:

> Less visible to users when a given setting is controlled by sysbuild

I like your idea of addressing this by borrowing from #55616 and generating `prompt`/`help` adjustments.

In addition, I would like to see more verbose warnings/errors printed by CMake when I contradict a sysbuild-controlled value using menuconfig. I'm not satisfied with the small "set more than once" warning, which doesn't even always appear when I expect it to. It would also be nice to error out on something like this:
```
-DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DCONFIG_BOOTLOADER_MCUBOOT=n
```

That would be the next step, and I believe this is the more promising approach to build upon.

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -321,16 +342,25 @@ function(ExternalZephyrProject_Cmake)
     COMMAND ${CMAKE_COMMAND}
       -G${CMAKE_GENERATOR}
         ${CMAKE_ARGS}
+      -DCONFIG_FORCED:FILEPATH=${dotconfigsysbuild}
```
> * Dependency on `${BINARY_DIR}/zephyr/.config`, to re-run sysbuild when image configuration changes.
 
I understand you reasons but extreme care must be taken, cause remember a CMake re-run of sysbuild will trigger a CMake re-run of the image which might result in an updated image `.config`, which then again may trigger a sysbuild CMake re-run if such a dependency is created.
So a risk of each a full CMake re-run before each build.

> * Consistency checks on sysbuild-controlled symbols, performed after `import_kconfig()`.
 
would be nice, but unsure how it will work out in practice, cause a user doing a `<image>_menuconfig` will only trigger a rebuild in that image (and the CMake of that image) and dependents of the image.

However, it might be that the image itself can compare the `.config` and forced `.config` and issue a better warning if there is a mismatch.

> * Only set `CONFIG_FORCED` if sysbuild's own `.config` has changed, so that we can try to report user errors instead of immediately correcting them.

CONFIG_FORCED points to a file, not specific Kconfig setting.
Also it is a cache variable, so wouldn't change anything if it is only provided on first CMake run, or if we provide it on each invocation.

Secondly, we shouldn't report errors and expecting users to manually fix them.
First because they would anyway need to set the value to the CMake enforced value.
Second, because a FATAL_ERROR in CMake prevents users from running menuconfig, hence making it harder to fix such issues, ref: https://github.com/zephyrproject-rtos/zephyr/issues/9573

57300 replied:
> So a risk of each a full CMake re-run before each build.

That's a good point. So far, I haven't come across this, but I see that it's not preventable, so I retract suggestion (1).

> a user doing a <image>_menuconfig will only trigger a rebuild in that image (and the CMake of that image) and dependents of the image.

That's why I suggested (1) initially, so that sysbuild gets rebuilt as well. Without that, you're right, (2) won't work.

> However, it might be that the image itself can compare the .config and forced .config and issue a better warning if there is a mismatch.

I think that could work too. I assume this would have to be implemented in `kconfig.py`.

> CONFIG_FORCED points to a file, not specific Kconfig setting.

I know, but now that you point it out, it should be renamed to avoid confusing other people.

> Also it is a cache variable, so wouldn't change anything if it is only provided on first CMake run, or if we provide it on each invocation.

Technically, we should be able to pass `-UCONFIG_FORCED` if we want to unset the cache variable, right?

> FATAL_ERROR in CMake prevents users from running menuconfig

True. I've seen this issue and it is an annoyance. We should still print a warning before applying the forced value.

tejlmand replied:
> > > CONFIG_FORCED points to a file, not specific Kconfig setting.
> > 
> > 
> > I know, but now that you point it out, it should be renamed to avoid confusing other people.
> 
> that a very good suggestion. Maybe `FORCED_CONF_FILE`, to be aligned with `CONF_FILE` and `EXTRA_CONF_FILE` (see #57994).

Updated to `FORCED_CONF_FILE`.

57300 replied:
> Guess you missed this comment in top of the PR:

Ah no, that was just poor choice of words on my part. What I had in mind was to print a _better_ warning, as I suggested earlier, since I found the one given by kconfiglib to be a bit small and not self-explanatory in this context.

tejlmand replied:
>  What I had in mind was to print a better warning, as I suggested earlier, since I found the one given by kconfiglib to be a bit small and not self-explanatory in this context.

then we are aligned.
Agreed that a better warning would be nice, but we need to be a bit careful cause in principle we want to minimize customization to kconfiglib.py itself. Zephyr specifics should ideally be handled in `kconfig.py`.

Therefore I prefer that we look into improved message as followup to this PR.

57300 replied:
> Updated to `FORCED_CONF_FILE`.

Nice! Please update the commit message as well.

tejlmand replied:
> Nice! Please update the commit message as well.

Done.

nordicjm commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -224,7 +238,11 @@ function(ExternalZephyrProject_Add)
+  set_target_properties(${ZBUILD_APPLICATION} PROPERTIES KCONFIG_BINARY_DIR
+                        ${application_binary_dir}/Kconfig
+  )
   if(ZBUILD_MAIN_APP)
```
Unresolved

tejlmand replied:
Fixed

nordicjm commented on share/sysbuild/image_config.cmake:
```diff
@@ -0,0 +1,23 @@
+get_target_property(image_board ${image} BOARD)
+if((NOT image_board) OR ("${image_BOARD}" STREQUAL "${BOARD}"))
+  get_target_property(${image}_APP_TYPE ${image} APP_TYPE)
+  if(NOT ${image}_APP_TYPE STREQUAL BOOTLOADER)
```
```suggestion
  if(NOT ${image}_APP_TYPE STREQUAL "BOOTLOADER")
```

tejlmand replied:
done

nordicjm commented on share/sysbuild/image_config.cmake:
```diff
@@ -0,0 +1,23 @@
+    )
+  endif()
+
+  if("${SB_CONFIG_SIGNATURE_TYPE}" STREQUAL "NONE")
+    set_config_bool(${image} CONFIG_MCUBOOT_GENERATE_UNSIGNED_IMAGE y)
+  else()
+    set_config_bool(${image} CONFIG_MCUBOOT_GENERATE_UNSIGNED_IMAGE n)
+  endif()
```
Does this need an if() for checking if mcuboot is enabled?

tejlmand replied:
done

tejlmand committed:
sysbuild: board and board revision improvement for images

This commit improves the BOARD handling for Zephyr images.

Concatenate sysbuild BOARD and BOARD_REVISION to BOARD when creating the
image sysbuild cache file. This ensures that the proper sysbuild
BOARD@REVISION can be correctly inherited.

Current solution creates image specific <image>_BOARD cache variables
whenever a board revision is in use, even if the image is expected to
inherit the BOARD@REVISION value from sysbuild.

This improvement means that only when a dedicated board or
board@revision is specified for an image, then an image sysbuild cache
entry will be created.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> This is still an issue (doesn't work for boards with a default revision).

I know, which is why I didn't mark it as fixed 😉 
When looking at it I discovered another flaw wrt. BOARD that I wanted to address first.
Part of it which you have also discovered, but there is also the possibility of a user specifying a dedicated BOARD for an image, such as `-DBOARD=myboard -Dmcuboot=myboard@<revision>` where user wants to build mcuboot for specific revision even though the main app could be targeting another revision.

In such a case, the revision will be part of the image `BOARD` setting.

57300 commented on cmake/modules/dts.cmake:
```diff
@@ -106,6 +106,7 @@ set(DTS_DEPS                    ${PROJECT_BINARY_DIR}/zephyr.dts.d)
 # This generates DT information needed by the Kconfig APIs.
 set(GEN_DRIVER_KCONFIG_SCRIPT   ${DT_SCRIPTS}/gen_driver_kconfig_dts.py)
 # Generated Kconfig symbols go here.
+set_ifndef(KCONFIG_BINARY_DIR   ${CMAKE_BINARY_DIR}/Kconfig)
```
I'm not sure if setting a default value for a variable like this is the right policy. Certain build system-wide variables should probably be centralized, rather than letting each module set its own default value. That could lead to its own brand of subtle bugs in the long term.

There's also this comment further up, which already suggests that `KCONFIG_BINARY_DIR` is expected to be defined before this module is loaded:

```
# Required variables:
# - BINARY_DIR_INCLUDE_GENERATED: where to put generated include files
# - KCONFIG_BINARY_DIR: where to put generated Kconfig files
```

Would it be better to do the following instead? (in a separate PR)
- Add `include(generated_file_directories)` and `include(zephyr_module)`.
- Check if every required variable is defined; raise an error if not.

tejlmand replied:
> I'm not sure if setting a default value for a variable like this is the right policy.

Correct, it's not ideal, but improving is part of a longer term goal, as also mentioned here:
https://github.com/zephyrproject-rtos/zephyr/pull/55293#discussion_r1122798737

> There's also this comment further up, which already suggests that KCONFIG_BINARY_DIR is expected to be defined before this module is loaded:

which is also not ideal, generally speaking a module should be as self-contained as possible with proper handling of required and optional value, and sane defaults for parts which are not core for the module.

In this case, creation of Kconfig file should not be core to a dts module, cause one should be able to load just dts.cmake for devicetree creation without Kconfig if there is a need. (such as `cmake-package-helper filtering).

Therefore creation of a default value is better than to fail if not set.
But ideally the Kconfig files should probably not be created if the variable has not been defined.

> Would it be better to do the following instead? (in a separate PR)

Yes, I think so, I believe this might be left-over from #55616 where iirc it is actually needed.
Will try to move it out of this PR.

tejlmand replied:
Removed commits and code related to KCONFIG_BINARY_DIR from sysbuild.

57300 commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -311,6 +308,11 @@ function(ExternalZephyrProject_Cmake)
       list(APPEND sysbuild_cache_strings "${cache_entry}\n")
     endif()
   endforeach()
+  if(DEFINED BOARD_REVISION)
+    list(APPEND sysbuild_cache_strings "BOARD:STRING=${BOARD}@${BOARD_REVISION}\n")
+  else()
+    list(APPEND sysbuild_cache_strings "BOARD:STRING=${BOARD}\n")
+  endif()
```
This should be moved down and combined with:

```
if(${ZCMAKE_APPLICATION}_BOARD)
  ...
elseif(DEFINED BOARD_REVISION)
  ...
else()
  ...
endif()
```
Otherwise, the sysbuild cache may contain two BOARD values.

tejlmand replied:
> Otherwise, the sysbuild cache may contain two BOARD values.

No it will not.
It will contain a BOARD value which is the `GLOBAL` sysbuild value, and then this code:
https://github.com/zephyrproject-rtos/zephyr/blob/b8891f2fa5f4fba82dfe6222dbf8860213699be8/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L323-L327
will ensure a board entry in the form of `<image>_BOARD`, such as `mcuboot_BOARD` is added to the sysbuild image cache.

This ensures that this code:
https://github.com/zephyrproject-rtos/zephyr/blob/49c7aa56fa0c1e61de95577e5fdc1877872b549c/cmake/modules/extensions.cmake#L2555-L2560
will correctly pickup BOARD, if no dedicated `<image>_BOARD` is defined.

tejlmand replied:
as example, where a specific board is given for an image build:
```
$ cat build/blinky_sysbuild_cache.txt|grep BOARD
CACHED_BOARD:STRING=nrf9160dk_nrf9160
BOARD:STRING=nrf9160dk_nrf9160@0.7.0
blinky_BOARD:STRING=nrf52840dk_nrf52811

```

nordicjm reviewed (dismissed):
Needs some additional settings for mcuboot and applications (in 3.4) but I'll add those as a separate PR
